File: src/test/java/me/helloc/iot/tunnel/MessageBufferTest.java
package me.helloc.iot.tunnel;

import org.eclipse.paho.client.mqttv3.MqttAsyncClient;
import org.eclipse.paho.client.mqttv3.MqttCallbackExtended;
import org.eclipse.paho.client.mqttv3.MqttMessage;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;

import java.util.Map;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

class MessageBufferTest {
    @Test
    void messageArrivedStoresPayloadToBuffer() throws Exception {
        MqttAsyncClient client = mock(MqttAsyncClient.class);
        ScheduledExecutorService scheduler = mock(ScheduledExecutorService.class);
        when(scheduler.schedule(any(Runnable.class), anyLong(), any()))
                .thenReturn(mock(ScheduledFuture.class));

        MqttBufferedSubscriber manager = MqttBufferedSubscriber.builder()
                .brokerUrl("tcp://localhost:1883")
                .clientSupplier(() -> client)
                .scheduler(scheduler)
                .build();

        MqttBufferedSubscriber.MessageListener msgListener = mock(MqttBufferedSubscriber.MessageListener.class);
        manager.addMessageListener(msgListener);

        ArgumentCaptor<MqttCallbackExtended> captor = ArgumentCaptor.forClass(MqttCallbackExtended.class);
        verify(client).setCallback(captor.capture());

        MqttMessage message = new MqttMessage("data".getBytes());
        captor.getValue().messageArrived("test/topic", message);

        verify(msgListener).onMessageReceived("test/topic", "data");
        verify(msgListener).onBufferedBefore("test/topic", "data");
        verify(msgListener).onBufferedAfter("test/topic", "data");

        Map.Entry<String, String> pair = manager.messageBuffer.poll();
        assertNotNull(pair);
        assertEquals("test/topic", pair.getKey());
        assertEquals("data", pair.getValue());
    }
}

File: src/test/kotlin/me/helloc/iot/tunnel/MessageBufferTest.kt
package me.helloc.iot.tunnel

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import org.eclipse.paho.client.mqttv3.MqttAsyncClient
import org.eclipse.paho.client.mqttv3.MqttCallbackExtended
import org.eclipse.paho.client.mqttv3.MqttMessage
import org.mockito.kotlin.argumentCaptor
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledFuture

class MessageBufferTest : StringSpec({
    "messageArrived stores payload to buffer" {
        val client = mock<MqttAsyncClient>()
        val scheduler = mock<ScheduledExecutorService>()
        whenever(scheduler.schedule(org.mockito.kotlin.any<Runnable>(), org.mockito.kotlin.any<Long>(), org.mockito.kotlin.any()))
            .thenReturn(mock<ScheduledFuture<*>>())

        val manager = MqttBufferedSubscriber.builder()
            .brokerUrl("tcp://localhost:1883")
            .clientSupplier { client }
            .scheduler(scheduler)
            .build()

        val msgListener = mock<MqttBufferedSubscriber.MessageListener>()
        manager.addMessageListener(msgListener)

        val captor = argumentCaptor<MqttCallbackExtended>()
        verify(client).setCallback(captor.capture())

        val message = MqttMessage("data".toByteArray())
        captor.firstValue.messageArrived("test/topic", message)

        verify(msgListener).onMessageReceived("test/topic", "data")
        verify(msgListener).onBufferedBefore("test/topic", "data")
        verify(msgListener).onBufferedAfter("test/topic", "data")

        manager.messageBuffer.poll() shouldBe ("test/topic" to "data")
    }
})
